File: 13-Stack-and-Queue/java/07-SortAStack.java
class Solution {
    public void sortStack(Stack<Integer> st) {
        if (st.isEmpty()) return;

        int top = st.pop();
        sortStack(st);
        insertSorted(st, top);
    }

    private void insertSorted(Stack<Integer> st, int val) {
        if (st.isEmpty() || st.peek() <= val) {
            st.push(val);
            return;
        }

        int top = st.pop();
        insertSorted(st, val);
        st.push(top);
    }
}

File: 13-Stack-and-Queue/python/07_sort_a_stack.py
class Solution:
    def sortStack(self, st: List[int]) -> None:
        if not st:
            return

        top = st.pop()
        self.sortStack(st)
        self._insert_sorted(st, top)

    def _insert_sorted(self, st: List[int], val: int) -> None:
        if not st or st[-1] <= val:
            st.append(val)
            return

        top = st.pop()
        self._insert_sorted(st, val)
        st.append(top)
